app/validators/answer_code_validator.py

```diff
@@ -0,0 +1,197 @@
+        self.validate_data_version()
+        self.validate_duplicate_answer_codes()
+        self.validate_missing_answer_id()
+        self.validate_missing_answer_codes()
+        self.validate_answer_codes_at_option_level()
+        self.validate_dynamic_options()
+        return self.errors
+
+    def validate_data_version(self):
+        if self.data_version != "0.0.3":
+            self.add_error(self.INCORRECT_DATA_VERSION_FOR_ANSWER_CODES)
+
+    def validate_duplicate_answer_codes(self):
+        duplicates = find_duplicates(self.codes)
+
+        if len(duplicates) > 0:
+            self.add_error(self.DUPLICATE_ANSWER_CODE_FOUND, duplicates=duplicates)
+
+    def validate_missing_answer_id(self):
+        for answer_code_id in self.answer_codes_answer_ids:
+            if answer_code_id not in self.all_answer_ids:
+                self.add_error(
+                    self.ANSWER_CODE_ANSWER_ID_NOT_FOUND_IN_SCHEMA,
+                    **{"answer_codes.answer_id": answer_code_id},
+                )
+
+    def validate_missing_answer_codes(self):
+        for answer_id in self.all_answer_ids:
+            block = self.questionnaire_schema.get_block_by_answer_id(answer_id)
+
+            if block["type"] in ["ListEditQuestion", "ListRemoveQuestion"]:
+                if answer_id in self.answer_codes_answer_ids:
+                    self.add_error(
+                        self.INVALID_ANSWER_CODE_FOR_LIST_COLLECTOR, answer_id=answer_id
+                    )
+                break
+
+            if answer_id not in self.answer_codes_answer_ids:
+                self.add_error(self.MISSING_ANSWER_CODE, answer_id=answer_id)
+
+    def validate_dynamic_options(self):
+        for answer_id in self.questionnaire_schema.answers_with_context:
+            answer = self.questionnaire_schema.answers_with_context[answer_id]
+
+            if "dynamic_options" in answer["answer"]:
+                answer_codes_for_options = [
+                    answer_code
+                    for answer_code in self.answer_codes
+                    if answer_code["answer_id"] == answer_id
+                ]
+
+                top_level_answer_code_count = sum(
+                    "answer_value" not in answer_code
+                    for answer_code in answer_codes_for_options
+                )
+                if top_level_answer_code_count == 0:
+                    self.add_error(
+                        self.DYNAMIC_ANSWER_OPTION_MUST_HAVE_ANSWER_CODE_SET_AT_TOP_LEVEL,
+                        answer_codes_for_options=answer_codes_for_options,
+                    )
+
+    def validate_answer_codes_at_option_level(self):
+        for answer_id in self.questionnaire_schema.answers_with_context:
+            answer = self.questionnaire_schema.answers_with_context[answer_id]
+
+            block = self.questionnaire_schema.get_block_by_answer_id(answer_id)
+
+            if block["type"] in ["ListEditQuestion", "ListRemoveQuestion"]:
```

Comment:
Need to handle primary person block since we support that in runner. They are named slightly differently for primary.